Classes/DDFileLogger.m

```diff
@@ -1531,3 +1540,231 @@ BOOL doesAppRunInBackground() {
+    }
+    return self;
+}
+
+#pragma mark - Subclass
+- (void)logData:(NSData *)data {
+    [self putDataIntoBufferOrToDisk:data];
+}
+
+#pragma mark - Initialization
+- (void)cleanup {
+    [self dumpBufferToDisk];
+    [super cleanup];
+}
+
+- (BOOL)isBufferFull {
+    return _bufferSize > _maximumBytesCountInBuffer;
+}
+
+- (void)putDataIntoBufferOrToDisk:(NSData *)data {
+    if ([self isBufferFull]) {
+        [self dumpBufferToDisk];
+    }
+    [self appendToBuffer:data];
+}
+
+@end
+
+@implementation DDFileLogger (ClassCluster)
+
++ (instancetype)createLoggerWithBuffer {
+    return [DDFileLoggerWithBuffer new];
+}
+
+@end
+
+// MARK: Public Interface
+@interface DDBufferedProxy: NSProxy
++ (instancetype)decoratedInstance:(DDFileLogger *)instance;
+@property (assign, nonatomic, readwrite) NSUInteger maximumBytesCountInBuffer;
+@end
+
+@interface DDBufferedProxy () {
+    NSOutputStream *_bufferStream;
+    NSUInteger _bufferSize;
+}
+- (instancetype)initWithInstance:(DDFileLogger *)instance;
+@property (strong, nonatomic, readwrite) DDFileLogger* instance;
+@end
+
+@interface DDBufferedProxy (StreamManipulation)
+- (void)flushBuffer;
+- (void)dumpBufferToDisk;
+- (void)appendToBuffer:(NSData *)data;
+- (BOOL)isBufferFull;
+@end
+
+@implementation DDBufferedProxy (StreamManipulation)
+
+- (void)flushBuffer {
+    // do something.
+    [_bufferStream close];
+    _bufferStream = nil;
+    _bufferSize = 0;
+}
+
+- (void)dumpBufferToDisk {
+    // do something.
+    __auto_type data = (NSData *)[_bufferStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
+    [self.instance logData:data];
+    [self flushBuffer];
+}
+
+- (void)appendToBuffer:(NSData *)data {
+    __auto_type length = data.length;
+    if (data.length != 0) {
+        if (_bufferStream == nil) {
+            _bufferStream = [[NSOutputStream alloc] initToMemory];
+            [_bufferStream open];
+            _bufferSize = 0;
+        }
+        const uint8_t *appendedData = calloc(length, sizeof(uint8_t));
+        [data getBytes:(void *)appendedData length:length];
+        if (appendedData != NULL) {
+            [_bufferStream write:appendedData maxLength:length];
+        }
+        if (appendedData != NULL) {
+            free((void *)appendedData);
+        }
+        _bufferSize += _bufferSize + length;
+    }
+}
+
+- (BOOL)isBufferFull {
+    return _bufferSize > self.maximumBytesCountInBuffer;
+}
+
+@end
+
+@implementation DDBufferedProxy
+@synthesize maximumBytesCountInBuffer = _maximumBytesCountInBuffer;
+
+#pragma mark - Initialization
++ (instancetype)decoratedInstance:(DDFileLogger *)instance {
+    return [[self alloc] initWithInstance:instance];
+}
+
+- (instancetype)initWithInstance:(DDFileLogger *)instance {
+    self.instance = instance;
+    self.maximumBytesCountInBuffer = kDefaultBytesCountInBuffer;
+    return self;
+}
+
+- (void)dealloc {
+    [self dumpBufferToDisk];
+    self.instance = nil;
```

Comment:
This isn't necessary, right? All properties will be set to `nil` on dealloc anyways.

Reply:
@ffried 
Sure, it is an explicit assignment to specify order of dealloc.
It could be removed.